From 42c6d8428c22b165523a75dde31a5bc98d554d02 Mon Sep 17 00:00:00 2001
From: Shinobu Uehara <shinobu.uehara.xc@renesas.com>
Date: Thu, 27 Jun 2013 19:43:13 +0900
Subject: [PATCH 697/715] ARM: shmobile: r8a7790: Add VIN driver support

Signed-off-by: Shinobu Uehara <shinobu.uehara.xc@renesas.com>
(cherry picked from commit 825a706183b4c19ff876515970caeb109c063e84)

Conflicts:
	arch/arm/mach-shmobile/board-lager.c
	arch/arm/mach-shmobile/setup-r8a7790.c

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 arch/arm/configs/shmobile_defconfig    |    4 +
 arch/arm/mach-shmobile/board-lager.c   |   22 ++++
 arch/arm/mach-shmobile/clock-r8a7790.c |    9 ++
 arch/arm/mach-shmobile/setup-r8a7790.c |  200 +++++++++++++++++++++++++++++++-
 4 files changed, 234 insertions(+), 1 deletion(-)

diff --git a/arch/arm/configs/shmobile_defconfig b/arch/arm/configs/shmobile_defconfig
index fc0aff3..bddb49e 100644
--- a/arch/arm/configs/shmobile_defconfig
+++ b/arch/arm/configs/shmobile_defconfig
@@ -130,6 +130,10 @@ CONFIG_SPI_SPIDEV=y
 # CONFIG_HWMON is not set
 CONFIG_REGULATOR=y
 CONFIG_MEDIA_SUPPORT=y
+CONFIG_VIDEO_DEV=y
+CONFIG_VIDEO_ADV7180=y
+CONFIG_VIDEO_ADV7612=y
+CONFIG_VIDEO_VIN=y
 CONFIG_MEDIA_CAMERA_SUPPORT=y
 CONFIG_V4L_PLATFORM_DRIVERS=y
 CONFIG_SOC_CAMERA=y
diff --git a/arch/arm/mach-shmobile/board-lager.c b/arch/arm/mach-shmobile/board-lager.c
index f1d6b7b..d43357f 100644
--- a/arch/arm/mach-shmobile/board-lager.c
+++ b/arch/arm/mach-shmobile/board-lager.c
@@ -163,6 +163,28 @@ static const struct pinctrl_map lager_pinctrl_map[] = {
 				  "msiof1_ctrl", "msiof1"),
 	PIN_MAP_MUX_GROUP_DEFAULT("spi_sh_msiof.1", "pfc-r8a7790",
 				  "msiof1_data", "msiof1"),
+	/* VIN0 */
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_data_g", "vin0"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_data_r", "vin0"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_data_b", "vin0"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_hsync_signal", "vin0"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_vsync_signal", "vin0"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_field_signal", "vin0"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_data_enable", "vin0"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.0", "pfc-r8a7790",
+				  "vin0_clk", "vin0"),
+	/* VIN1 */
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.1", "pfc-r8a7790",
+				  "vin1_data", "vin1"),
+	PIN_MAP_MUX_GROUP_DEFAULT("vin.1", "pfc-r8a7790",
+				  "vin1_clk", "vin1"),
 };
 
 static void __init lager_add_standard_devices(void)
diff --git a/arch/arm/mach-shmobile/clock-r8a7790.c b/arch/arm/mach-shmobile/clock-r8a7790.c
index f482328..b43f3e3 100644
--- a/arch/arm/mach-shmobile/clock-r8a7790.c
+++ b/arch/arm/mach-shmobile/clock-r8a7790.c
@@ -205,6 +205,7 @@ enum {
 	MSTP112,
 	MSTP219, MSTP218,
 	MSTP502, MSTP501,
+	MSTP811, MSTP810, MSTP809, MSTP808,
 	MSTP704, MSTP703,
 	MSTP815, MSTP814,
 	MSTP917, MSTP931, MSTP930, MSTP929, MSTP928, MSTP922,
@@ -253,6 +254,10 @@ static struct clk mstp_clks[MSTP_NR] = {
 	[MSTP929] = SH_CLK_MSTP32(&hp_clk, SMSTPCR9, 29, 0),
 	[MSTP928] = SH_CLK_MSTP32(&hp_clk, SMSTPCR9, 28, 0),
 	[MSTP922] = SH_CLK_MSTP32(&hp_clk, SMSTPCR9, 22, 0),
+	[MSTP811] = SH_CLK_MSTP32(&zg_clk, SMSTPCR8, 11, 0),
+	[MSTP810] = SH_CLK_MSTP32(&zg_clk, SMSTPCR8, 10, 0),
+	[MSTP809] = SH_CLK_MSTP32(&zg_clk, SMSTPCR8,  9, 0),
+	[MSTP808] = SH_CLK_MSTP32(&zg_clk, SMSTPCR8,  8, 0),
 	[MSTP1031] = SH_CLK_MSTP32(&hp_clk, SMSTPCR10, 31, 0),
 	[MSTP1030] = SH_CLK_MSTP32(&hp_clk, SMSTPCR10, 30, 0),
 	[MSTP1019] = SH_CLK_MSTP32(&hp_clk, SMSTPCR10, 19, 0),
@@ -347,6 +352,10 @@ static struct clk_lookup lookups[] = {
 	CLKDEV_DEV_ID("i2c-rcar.2", &mstp_clks[MSTP929]),
 	CLKDEV_DEV_ID("i2c-rcar.3", &mstp_clks[MSTP928]),
 	CLKDEV_CON_ID("adg", &mstp_clks[MSTP922]),
+	CLKDEV_DEV_ID("vin.0", &mstp_clks[MSTP811]),
+	CLKDEV_DEV_ID("vin.1", &mstp_clks[MSTP810]),
+	CLKDEV_DEV_ID("vin.2", &mstp_clks[MSTP809]),
+	CLKDEV_DEV_ID("vin.3", &mstp_clks[MSTP808]),
 	CLKDEV_CON_ID("src0", &mstp_clks[MSTP1031]),
 	CLKDEV_CON_ID("src1", &mstp_clks[MSTP1030]),
 	CLKDEV_CON_ID("dvc0", &mstp_clks[MSTP1019]),
diff --git a/arch/arm/mach-shmobile/setup-r8a7790.c b/arch/arm/mach-shmobile/setup-r8a7790.c
index ca5e85b..069d859 100644
--- a/arch/arm/mach-shmobile/setup-r8a7790.c
+++ b/arch/arm/mach-shmobile/setup-r8a7790.c
@@ -29,6 +29,7 @@
 #include <linux/sh_eth.h>
 #include <linux/i2c/i2c-rcar.h>
 #include <linux/sh_dma-desc.h>
+#include <linux/dma-mapping.h>
 #include <linux/sh_audma-pp.h>
 #include <linux/platform_data/gpio-rcar.h>
 #include <linux/platform_data/rcar-du.h>
@@ -48,6 +49,7 @@
 #include <linux/spi/flash.h>
 #include <linux/dma-mapping.h>
 #include <linux/spi/sh_msiof.h>
+#include <media/vin.h>
 #include <mach/common.h>
 #include <mach/irqs.h>
 #include <mach/r8a7790.h>
@@ -1669,6 +1671,196 @@ static struct platform_device mmc1_device = {
 	}
 };
 
+/* VIN */
+static struct resource vin0_resources[] = {
+	[0] = {
+		.name = "VIN0",
+		.start = 0xe6ef0000,
+		.end = 0xe6ef0fff,
+		.flags = IORESOURCE_MEM,
+	},
+	[1] = {
+		.start = gic_spi(188),
+		.flags = IORESOURCE_IRQ,
+	},
+};
+
+static struct resource vin1_resources[] = {
+	[0] = {
+		.name = "VIN1",
+		.start = 0xe6ef1000,
+		.end = 0xe6ef1fff,
+		.flags = IORESOURCE_MEM,
+	},
+	[1] = {
+		.start = gic_spi(189),
+		.flags = IORESOURCE_IRQ,
+	},
+};
+
+static struct resource vin2_resources[] = {
+	[0] = {
+		.name = "VIN2",
+		.start = 0xe6ef2000,
+		.end = 0xe6ef2fff,
+		.flags = IORESOURCE_MEM,
+	},
+	[1] = {
+		.start = gic_spi(190),
+		.flags = IORESOURCE_IRQ,
+	},
+};
+
+static struct resource vin3_resources[] = {
+	[0] = {
+		.name = "VIN3",
+		.start = 0xe6ef3000,
+		.end = 0xe6ef3fff,
+		.flags = IORESOURCE_MEM,
+	},
+	[1] = {
+		.start = gic_spi(191),
+		.flags = IORESOURCE_IRQ,
+	},
+};
+
+static struct vin_info vin_info[] = {
+	[0] = {
+		.input = VIN_INPUT_ITUR_BT709_24BIT,
+		.flags = 0,
+	},
+	[1] = {
+		.input = VIN_INPUT_ITUR_BT656_8BIT,
+		.flags = 0,
+	},
+	[2] = {
+		.input = VIN_INPUT_UNDEFINED,
+		.flags = 0,
+	},
+	[3] = {
+		.input = VIN_INPUT_UNDEFINED,
+		.flags = 0,
+	},
+};
+
+static u64 vin_dmamask = DMA_BIT_MASK(32);
+
+static struct platform_device vin0_device = {
+	.name  = "vin",
+	.id = 0,
+	.num_resources = ARRAY_SIZE(vin0_resources),
+	.resource  = vin0_resources,
+	.dev  = {
+		.dma_mask = &vin_dmamask,
+		.platform_data = &vin_info[0],
+		.coherent_dma_mask = DMA_BIT_MASK(32),
+	},
+};
+
+static struct platform_device vin1_device = {
+	.name  = "vin",
+	.id = 1,
+	.num_resources = ARRAY_SIZE(vin1_resources),
+	.resource  = vin1_resources,
+	.dev  = {
+		.dma_mask = &vin_dmamask,
+		.platform_data = &vin_info[1],
+		.coherent_dma_mask = DMA_BIT_MASK(32),
+	},
+};
+
+static struct platform_device vin2_device = {
+	.name  = "vin",
+	.id = 2,
+	.num_resources = ARRAY_SIZE(vin2_resources),
+	.resource  = vin2_resources,
+	.dev  = {
+		.dma_mask = &vin_dmamask,
+		.platform_data = &vin_info[2],
+		.coherent_dma_mask = DMA_BIT_MASK(32),
+	},
+};
+
+static struct platform_device vin3_device = {
+	.name  = "vin",
+	.id = 3,
+	.num_resources = ARRAY_SIZE(vin3_resources),
+	.resource  = vin3_resources,
+	.dev  = {
+		.dma_mask = &vin_dmamask,
+		.platform_data = &vin_info[3],
+		.coherent_dma_mask = DMA_BIT_MASK(32),
+	},
+};
+
+static struct i2c_board_info lager_i2c_camera[] = {
+	{ I2C_BOARD_INFO("adv7612", 0x4C), },
+	{ I2C_BOARD_INFO("adv7180", 0x20), },
+};
+
+static void camera_power_on(void)
+{
+	return;
+}
+
+static void camera_power_off(void)
+{
+	return;
+}
+
+static int adv7612_power(struct device *dev, int mode)
+{
+	if (mode)
+		camera_power_on();
+	else
+		camera_power_off();
+
+	return 0;
+}
+
+static int adv7180_power(struct device *dev, int mode)
+{
+	if (mode)
+		camera_power_on();
+	else
+		camera_power_off();
+
+	return 0;
+}
+
+static struct soc_camera_link adv7612_ch0_link = {
+	.bus_id = 0,
+	.power  = adv7612_power,
+	.board_info = &lager_i2c_camera[0],
+	.i2c_adapter_id = 2,
+	.module_name = "adv7612",
+};
+
+static struct soc_camera_link adv7180_ch1_link = {
+	.bus_id = 1,
+	.power  = adv7180_power,
+	.board_info = &lager_i2c_camera[1],
+	.i2c_adapter_id = 2,
+	.module_name = "adv7180",
+};
+
+static struct platform_device soc_camera_device[] = {
+	{
+		.name = "soc-camera-pdrv",
+		.id = 0,
+		.dev = {
+			.platform_data = &adv7612_ch0_link,
+		},
+	},
+	{
+		.name = "soc-camera-pdrv",
+		.id = 1,
+		.dev = {
+			 .platform_data = &adv7180_ch1_link,
+		},
+	},
+};
+
 static struct platform_device *r8a7790_early_devices[] __initdata = {
 	&eth_device,
 	&powervr_device,
@@ -1706,6 +1898,12 @@ static struct platform_device *r8a7790_early_devices[] __initdata = {
 	&sata1_device,
 	&mmc0_device,
 	&mmc1_device,
+	&vin0_device,
+	&vin1_device,
+	&vin2_device,
+	&vin3_device,
+	&soc_camera_device[0],
+	&soc_camera_device[1],
 };
 
 static struct renesas_irqc_config irqc0_data = {
@@ -1783,7 +1981,7 @@ void __init r8a7790_timer_init(void)
 }
 
 struct sys_timer r8a7790_timer = {
-        .init           = r8a7790_timer_init,
+	.init		= r8a7790_timer_init,
 };
 
 #ifdef CONFIG_USE_OF
-- 
1.7.10.4

